go/merkletree2/tree_test.go

```diff
@@ -0,0 +1,417 @@
+
+				// empty proof
+				err = verifier.VerifyInclusionProof(context.TODO(), kvp, MerkleInclusionProof{}, rootHash1)
+				require.Error(t, err)
+				require.IsType(t, ProofVerificationFailedError{}, err)
+
+				// Change the blinding key-specific secret (where appropriate)
+				if tree.cfg.useBlindedValueHashes {
+					require.NotNil(t, proof.KeySpecificSecret)
+					require.True(t, len(proof.KeySpecificSecret) > 0, "Kss: %X", proof.KeySpecificSecret)
+
+					fakeKSS := KeySpecificSecret(append([]byte(nil), ([]byte)(proof.KeySpecificSecret)...))
+					([]byte(fakeKSS))[0] = 1 + ([]byte(fakeKSS))[0]
+					var fakeProof MerkleInclusionProof
+					fakeProof = proof
+					fakeProof.KeySpecificSecret = fakeKSS
+					err = verifier.VerifyInclusionProof(context.TODO(), kvp, fakeProof, rootHash1)
+					require.Error(t, err)
+					require.IsType(t, ProofVerificationFailedError{}, err)
+				}
+			}
+		})
+	}
+
+}
+
+func TestGetLatestRoot(t *testing.T) {
+	config1bitU, config2bitsU, config3bitsU := getTreeCfgsWith1_2_3BitsPerIndexUnblinded(t)
+	config1bitB, config2bitsB, config3bitsB := getTreeCfgsWith1_2_3BitsPerIndexBlinded(t)
+	kvps1_1bit, kvps2_1bit, kvps3_1bit := getSampleKVPS1bit()
+	kvps1_3bits, kvps2_3bits, kvps3_3bits := getSampleKVPS3bits()
+
+	config3bits2valsPerLeafU, err := NewConfig(IdentityHasher{}, false, 3, 2, 3)
+	require.NoError(t, err)
+	config3bits2valsPerLeafB, err := NewConfig(IdentityHasherBlinded{}, true, 3, 2, 3)
+	require.NoError(t, err)
+
+	tests := []struct {
+		cfg   Config
+		kvps1 []KeyValuePair
+		kvps2 []KeyValuePair
+		kvps3 []KeyValuePair
+	}{
+		{config1bitU, kvps1_1bit, kvps2_1bit, kvps3_1bit},
+		{config2bitsU, kvps1_1bit, kvps2_1bit, kvps3_1bit},
+		{config3bitsU, kvps1_3bits, kvps2_3bits, kvps3_3bits},
+		{config3bits2valsPerLeafU, kvps1_3bits, kvps2_3bits, kvps3_3bits},
+		{config1bitB, kvps1_1bit, kvps2_1bit, kvps3_1bit},
+		{config2bitsB, kvps1_1bit, kvps2_1bit, kvps3_1bit},
+		{config3bitsB, kvps1_3bits, kvps2_3bits, kvps3_3bits},
+		{config3bits2valsPerLeafB, kvps1_3bits, kvps2_3bits, kvps3_3bits},
+	}
+
+	for _, test := range tests {
+		t.Run(fmt.Sprintf("%v bits %v values per leaf tree (blinded %v)", test.cfg.bitsPerIndex, test.cfg.maxValuesPerLeaf, test.cfg.useBlindedValueHashes), func(t *testing.T) {
+			tree, err := NewTree(test.cfg, &InMemoryStorageEngine{}, logger.NewTestLogger(t))
+			require.NoError(t, err)
+
+			rootHash1Exp, err := tree.Build(context.TODO(), nil, test.kvps1)
+			require.NoError(t, err)
+
+			_, _, rootHash1, err := tree.GetLatestRoot(context.TODO(), nil)
+			require.NoError(t, err)
+			require.True(t, rootHash1Exp.Equal(rootHash1))
+
+			rootHash2Exp, err := tree.Build(context.TODO(), nil, test.kvps2)
+			require.NoError(t, err)
+
+			_, _, rootHash2, err := tree.GetLatestRoot(context.TODO(), nil)
+			require.NoError(t, err)
+			require.True(t, rootHash2Exp.Equal(rootHash2))
+
+			require.False(t, rootHash1.Equal(rootHash2))
+		})
+	}
+
+}
```

Comment:
As noted in person, need more tests for bigger deeper trees. Can use random keys.

Reply:
For instance, make a test of a binary and 4-way tree that each have 10000 values in them.